### ISSUE TITLE
Transaction receipt events not getting decoded properly

### ISSUE THREAD

**Issue by gakonst** [ethereum/web3.py#503]
* Version: 4.0.0b4
* Python: 3.5 and 3.6
* OS: Ubuntu 17.04

### What was wrong?
Expected for the get_new_entries() method to catch the event. Instead it seems to not be able to decode the transaction receipt. I have tested the error with ethereumjs-testrpc (ganache-cli), geth --dev mode, full node running on private chain (custom genesis.json) and with eth-tester

### Installed python packages (output of pip freeze after installing pyethereum/web3 in fresh venv)
```
bitcoin==1.1.42
certifi==2017.11.5
cffi==1.11.2
chardet==3.0.4
cytoolz==0.8.2
eth-abi==0.5.0
eth-keyfile==0.4.0
eth-keys==0.1.0b3
eth-tester==0.1.0b5
eth-utils==0.7.1
ethereum==1.6.1
idna==2.6
lru-dict==1.1.6
pbkdf2==1.3
py-solc==2.1.0
pycparser==2.18
pycryptodome==3.4.7
pyethash==0.1.27
pysha3==1.0.2
PyYAML==3.12
repoze.lru==0.7
requests==2.18.4
rlp==0.6.0
scrypt==0.8.0
secp256k1==0.13.2
semantic-version==2.6.0
toolz==0.8.2
typing==3.6.2
urllib3==1.22
web3==4.0.0b4
```

## Script Output with error
I print the ABI first, just to make sure it's getting encoded properly. Then, the get_all_entries() function returns an empty list, as expected. After making a call that should fire the event, I get a proper transaction receipt, however the get_new_entries() function is unable to decode it and return a populated list.

```py
[{'constant': False, 'inputs': [], 'name': 'test', 'outputs': [], 'payable': False, 'stateMutability': 'nonpayable', 'type': 'function'}, {'anonymous': False, 'inputs': [{'indexed': True, 'name': 'x', 'type': 'uint256'}], 'name': 'Test', 'type': 'event'}]
[]
AttributeDict({'transactionHash': HexBytes('0x09a123fdc71d957d0929102a85a6aff36d89206970aad99518a5ce724b89bcbc'), 'transactionIndex': 0, 'blockNumber': 2, 'blockHash': HexBytes('0x37165d5ec1d15227c46b6a5364535f1b6e525d2ff9d047279b8880ad469c2da0'), 'cumulativeGasUsed': 22555, 'gasUsed': 22555, 'contractAddress': None, 'logs': [AttributeDict({'type': 'mined', 'logIndex': 0, 'transactionIndex': 0, 'transactionHash': HexBytes('0x09a123fdc71d957d0929102a85a6aff36d89206970aad99518a5ce724b89bcbc'), 'blockHash': HexBytes('0x37165d5ec1d15227c46b6a5364535f1b6e525d2ff9d047279b8880ad469c2da0'), 'blockNumber': 2, 'address': '0xc305c901078781C232A2a521C2aF7980f8385ee9', 'data': '0x', 'topics': [HexBytes('0x63a242a632efe33c0e210e04e4173612a17efa4f16aa4890bc7e46caece80de0'), HexBytes('0x0000000000000000000000000000000000000000000000000000000000000002')]})]})
/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/bitcoin/mnemonic.py:7: ResourceWarning: unclosed file <_io.TextIOWrapper name='/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/bitcoin/english.txt' mode='r' encoding='UTF-8'>
  wordlist_english=list(open(os.path.join(os.path.dirname(os.path.realpath(__file__)),'english.txt'),'r'))
/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/eth_abi/abi.py:92: DeprecationWarning: Automatic inference of hex encoded data has been deprecated. Please adjust your code to ensure that the data argument for `decode_single` is a byte string
  "Automatic inference of hex encoded data has been deprecated. "
Traceback (most recent call last):
  File "test.py", line 56, in <module>
    print (transfer_filter.get_new_entries())
  File "/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/web3/utils/filters.py", line 103, in get_new_entries
    return self._format_log_entries(log_entries)
  File "/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/web3/utils/filters.py", line 114, in _format_log_entries
    self.format_entry(log_entry) for log_entry in log_entries
  File "/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/web3/utils/filters.py", line 114, in <listcomp>
    self.format_entry(log_entry) for log_entry in log_entries
  File "/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/web3/utils/filters.py", line 159, in format_entry
    return self.log_entry_formatter(entry)
  File "/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/web3/utils/events.py", line 204, in get_event_data
    'logIndex': log_entry['logIndex'],
KeyError: 'logIndex'
```

### Code
```py
import web3
from web3 import Web3, HTTPProvider
from web3.providers.eth_tester import EthereumTesterProvider
from eth_tester import EthereumTester
from solc import compile_source

# Solidity source code
contract_source_code = '''
pragma solidity ^0.4.18;

contract C {
    event Test(uint indexed x);
    function test()  {
        Test(2);
    }
}
'''

compiled_sol = compile_source(contract_source_code) # Compiled source code
contract_interface = compiled_sol['<stdin>:C']
print (contract_interface['abi'])

# web3.py instance
eth_tester = EthereumTester()
web3 = Web3(EthereumTesterProvider(eth_tester))
accounts = eth_tester.get_accounts()
# web3 = Web3(HTTPProvider("http://localhost:8545"))
# accounts = web3.eth.accounts

# Instantiate and deploy contract
contract = web3.eth.contract(abi=contract_interface['abi'], bytecode=contract_interface['bin'])

# Get transaction hash from deployed contract
tx_hash = contract.deploy(transaction={'from': accounts[0], 'gas': 410000})

# Get tx receipt to get contract address
tx_receipt = web3.eth.getTransactionReceipt(tx_hash)
contract_address = tx_receipt['contractAddress']

# Contract instance 
contract_instance = web3.eth.contract(address=contract_address, abi=contract_interface['abi'])

transfer_filter = contract_instance.eventFilter('Test')

# Should be empty
print (transfer_filter.get_all_entries())
    
# Call function
tx_hash = contract_instance.transact({'from': accounts[0]}).test();
# Receipt's logs must have the event
receipt = web3.eth.getTransactionReceipt(tx_hash)
print (receipt)

# Should print the event
print (transfer_filter.get_new_entries())
```

### How can it be fixed?

Could be fixed by decoding the transaction properly, haven't thought of the implementation though.

**Comment by pipermerriam** [comment]
@gakonst would you be able to drop into a print statement into `/home/gak/.virtualenvs/web3bug/lib/python3.6/site-packages/web3/utils/filters.py` just before the return statement `return self.log_entry_formatter(entry)` that prints out `entry`.  I'm very curious what is being returned there.

**Comment by gakonst** [comment]
Managed to fix it, submitted a relevant pull request.
https://github.com/ethereum/web3.py/pull/504

**Comment by gakonst** [comment]
It looks like TestRPCProvider implements camelCase and EthereumTesterProvider implements snake_case. My changes fix the bug when using EthereumTesterProvider, however the tests in Travis use TestRPCProvider and so they fail. 
### Proof of Concept
```py
import web3
from web3 import Web3, HTTPProvider, TestRPCProvider
from web3.providers.eth_tester import EthereumTesterProvider
from eth_tester import EthereumTester
from solc import compile_source
import sys, traceback
# Solidity source code
contract_source_code = '''
pragma solidity ^0.4.18;

contract C {
    event Test(uint indexed x);
    function test()  {
        Test(2);
    }
}
'''

compiled_sol = compile_source(contract_source_code) # Compiled source code
contract_interface = compiled_sol['<stdin>:C']
print (contract_interface['abi'])

# web3.py instance
print("======================= ETH-TESTER =======================")
eth_tester = EthereumTester()
web3 = Web3(EthereumTesterProvider(eth_tester))
accounts = eth_tester.get_accounts()
contract = web3.eth.contract(abi=contract_interface['abi'], bytecode=contract_interface['bin'])
tx_hash = contract.deploy(transaction={'from': accounts[0], 'gas': 410000})
tx_receipt = web3.eth.getTransactionReceipt(tx_hash)
contract_address = tx_receipt['contractAddress']
contract_instance = web3.eth.contract(address=contract_address, abi=contract_interface['abi'])
transfer_filter = contract_instance.eventFilter('Test')
print (transfer_filter.get_all_entries())
tx_hash = contract_instance.transact({'from': accounts[0]}).test();
receipt = web3.eth.getTransactionReceipt(tx_hash)
print ('Tx receipt',receipt)
try :
    print (transfer_filter.get_new_entries())
except:
    print("Failed to get events")
    print ('-'*60)
    traceback.print_exc(file=sys.stdout)
    print ('-'*60)

print("======================= END =======================\n\n")


print("======================= ETH-TESTRPC =======================")
web3 = Web3(TestRPCProvider())
accounts = web3.eth.accounts
contract = web3.eth.contract(abi=contract_interface['abi'], bytecode=contract_interface['bin'])
tx_hash = contract.deploy(transaction={'from': accounts[0], 'gas': 410000})
tx_receipt = web3.eth.getTransactionReceipt(tx_hash)
contract_address = tx_receipt['contractAddress']
contract_instance = web3.eth.contract(address=contract_address, abi=contract_interface['abi'])
transfer_filter = contract_instance.eventFilter('Test')
print (transfer_filter.get_all_entries())
tx_hash = contract_instance.transact({'from': accounts[0]}).test();
receipt = web3.eth.getTransactionReceipt(tx_hash)
print ('Tx receipt',receipt)
try:
    print (transfer_filter.get_new_entries())
except:
    print("Failed to get events")
    print ('-'*60)
    traceback.print_exc(file=sys.stdout)
    print ('-'*60)
print("======================= END ======================")
```

**Comment by carver** [comment]
Fixed by #504 - will go in beta 5

**Comment by jerryji** [comment]
As of 21-Apr-2018, web3.py 4.1.0 python 3.6.3 under Ubuntu 17.10, I'm not getting any event (I'm only trying to print the last event from gakonst's code) --

```
$ python3 test.py 
test.py:34: DeprecationWarning: deploy is deprecated in favor of contract.constructor.transact
  tx_hash = contract.deploy(transaction={'from': accounts[0], 'gas': 410000})
/usr/local/lib/python3.6/dist-packages/web3/utils/decorators.py:14: DeprecationWarning: eventFilter is deprecated in favor of contract.events.<event name>.createFilter
  return self.method(obj, *args, **kwargs)
/usr/local/lib/python3.6/dist-packages/web3/utils/decorators.py:14: DeprecationWarning: transact is deprecated in favor of contract.<functions/events>.<method name>.transact
  return self.method(obj, *args, **kwargs)
[]
```

Sorry for my messy global pip3 freeze --

```
anyconfig==0.9.4
apturl==0.5.2
argh==0.26.2
asn1crypto==0.22.0
attrdict==2.0.0
attrs==17.4.0
beautifulsoup4==4.6.0
bitcoin==1.1.42
blinker==1.3
Brlapi==0.6.5
certifi==2018.4.16
cffi==1.11.5
chardet==3.0.4
chrome-gnome-shell==0.0.0
click==6.7
coincurve==7.1.0
command-not-found==0.3
contextlib2==0.5.5
cryptography==1.9
cupshelpers==1.0
cytoolz==0.9.0.1
decorator==4.1.1
defer==1.0.6
distro-info==0.17
eth-abi==1.0.0
eth-account==0.1.0a2
eth-hash==0.1.2
eth-keyfile==0.5.1
eth-keys==0.2.0b3
eth-rlp==0.1.0
eth-tester==0.1.0b23
eth-testrpc==1.3.3
eth-utils==1.0.3
ethereum==1.6.1
feedparser==5.1.3
future==0.16.0
hexbytes==0.1.0
html5lib==0.999999999
httplib2==0.9.2
idna==2.6
itsdangerous==0.24
json-rpc==1.10.8
jsonschema==2.6.0
kazam==1.4.5
keyring==10.4.0
keyrings.alt==2.2
language-selector==0.1
launchpadlib==1.10.5
lazr.restfulclient==0.13.5
lazr.uri==1.0.3
louis==3.0.0
lru-dict==1.1.6
meld==3.18.0
more-itertools==4.1.0
msgpack==0.5.4
msgpack-python==0.4.8
olefile==0.44
pathtools==0.1.2
pbkdf2==1.3
pexpect==4.2.1
pickleshare==0.7.4
Pillow==4.1.1
pluggy==0.6.0
populus==2.2.0
prompt-toolkit==1.0.14
psutil==5.0.1
py==1.5.3
py-ecc==1.4.2
py-geth==2.0.1
py-solc==2.1.0
pycparser==2.18
pycrypto==2.6.1
pycryptodome==3.6.1
pycups==1.9.73
pyethash==0.1.27
Pygments==2.2.0
pygobject==3.24.1
PyJWT==1.4.2
pylru==1.0.9
pysha3==1.0.2
pytest==3.5.0
python-apt==1.4.0b3
python-dateutil==2.7.1
python-debian==0.1.30
python-editor==1.0.3
python-xapp==1.0.0
python-xlib==0.14
pyxdg==0.25
PyYAML==3.12
reportlab==3.4.0
repoze.lru==0.7
requests==2.18.4
rlp==0.6.0
screen-resolution-extra==0.0.0
scrypt==0.8.6
secp256k1==0.13.2
SecretStorage==2.3.1
semantic-version==2.6.0
setproctitle==1.1.10
simplegeneric==0.8.1
simplejson==3.11.1
six==1.11.0
ssh-import-id==5.7
system-service==0.3
systemd-python==234
toolz==0.9.0
toposort==1.5
wadllib==1.3.2
Wand==0.4.4
watchdog==0.8.3
wcwidth==0.1.7
web3==4.1.0
webencodings==0.5

```
